kernel/objects/execution-context/objects/ExecutionContext.cc

```diff
@@ -510,35 +558,28 @@ namespace mythos {
         if (prevWait & IN_WAIT) {
             // clear IS_NOTIFIED only if the user mode was waiting for it
             // we won't clear it twice without a second wait() system call
-            clearFlags(IS_NOTIFIED); 
+            clearFlags(IS_NOTIFIED);
 
             // return a notification event if any
             auto e = notificationQueue.pull();
             if (e) {
                 auto ev = e->get()->deliver();
-                threadState.rsi = ev.user;
-                threadState.rdi = ev.state;
+                state->threadState.rsi = ev.user;
+                state->threadState.rdi = ev.state;
             } else {
-                threadState.rsi = 0;
-                threadState.rdi = uint64_t(Error::NO_MESSAGE);
+                state->threadState.rsi = 0;
+                state->threadState.rdi = uint64_t(Error::NO_MESSAGE);
             }
-            MLOG_DETAIL(mlog::ec, DVAR(this), "return one notification", 
-                        DVARhex(threadState.rsi), DVAR(threadState.rdi));
+            MLOG_DETAIL(mlog::ec, DVAR(this), "return one notification",
+                        DVARhex(state->threadState.rsi), DVAR(state->threadState.rdi));
         }
 
-        // Reload the address space if it has changed.
-        // It might have been revoked concurrently since we checked the blocking flags.
-        // In that case we abort the resume. The revoke will update NO_AS for us.
-        {
-            TypedCap<IPageMap> as(_as);
-            if (!as) return;
-            auto info = as->getPageMapInfo(as.cap());
-            MLOG_DETAIL(mlog::ec, "load addrspace", DVAR(this), DVARhex(info.table.physint()));
-            getLocalPlace().setCR3(info.table); // without reload if not changed
-        }
+        // Reload the address space and thread state address if they have changed
+        getLocalPlace().setCR3(page_table); // without reload if not changed
```

Comment:
We should be very careful never to set CR3 to `nullptr`. Maybe assert `page_table != nullptr` or at least use the default/fallback system PML4  as a "null" value.